File: apps/desktop/src/lib/tauri/wallet.ts
/**
 * Tauri IPC wrappers for wallet commands
 *
 * These functions wrap the Tauri invoke API for type-safe communication
 * with the Rust backend wallet module.
 */

import { invoke } from "@tauri-apps/api/core";

// =============================================================================
// Types matching Rust backend
// =============================================================================

export type ChainFamily = "secp256k1" | "ed25519" | "sr25519";

export interface ChainInfo {
  id: string;
  name: string;
  symbol: string;
  family: ChainFamily;
  coin_type: number;
  is_testnet: boolean;
  icon_name: string;
}

export interface DerivedAddress {
  chain: string;
  chain_family: ChainFamily;
  address: string;
  derivation_path: string;
  public_key: string; // hex encoded
}

export interface CreateHDWalletResponse {
  wallet_id: string;
  mnemonic: string;
  addresses: DerivedAddress[];
}

export interface ValidateMnemonicResponse {
  is_valid: boolean;
  word_count: number;
  error: string | null;
}

// =============================================================================
// Chain Information
// =============================================================================

/**
 * Get all supported blockchain chains
 */
export async function getSupportedChains(): Promise<ChainInfo[]> {
  return invoke<ChainInfo[]>("get_supported_chains");
}

/**
 * Get only mainnet chains (excludes testnets)
 */
export async function getMainnetChains(): Promise<ChainInfo[]> {
  return invoke<ChainInfo[]>("get_mainnet_chains");
}

/**
 * Validate an address for a specific chain
 */
export async function validateChainAddress(
  chainId: string,
  address: string
): Promise<boolean> {
  return invoke<boolean>("validate_chain_address", {
    chainId,
    address,
  });
}

// =============================================================================
// Mnemonic Operations
// =============================================================================

/**
 * Generate a new random mnemonic phrase
 * @param wordCount - Number of words (12 or 24)
 */
export async function generateMnemonic(wordCount: number = 12): Promise<string> {
  return invoke<string>("generate_mnemonic", { wordCount });
}

/**
 * Validate an existing mnemonic phrase
 */
export async function validateMnemonic(
  phrase: string
): Promise<ValidateMnemonicResponse> {
  return invoke<ValidateMnemonicResponse>("validate_mnemonic", { phrase });
}

// =============================================================================
// Wallet Creation
// =============================================================================

/**
 * Create a new HD wallet with generated mnemonic
 *
 * @param name - Display name for the wallet
 * @param chains - Chain IDs to derive addresses for
 * @param password - Password for encrypting the wallet
 * @param wordCount - Mnemonic word count (12 or 24)
 * @returns Wallet ID, mnemonic (for backup), and derived addresses
 *
 * @security The mnemonic is returned ONCE for backup purposes.
 * Display it for user to write down, then verify backup before proceeding.
 */
export async function createHDWallet(
  name: string,
  chains: string[],
  password: string,
  wordCount: number = 12
): Promise<CreateHDWalletResponse> {
  return invoke<CreateHDWalletResponse>("create_hd_wallet", {
    name,
    chains,
    password,
    wordCount,
  });
}

/**
 * Import an existing HD wallet from mnemonic
 */
export async function importHDWallet(
  name: string,
  mnemonic: string,
  chains: string[],
  password: string
): Promise<CreateHDWalletResponse> {
  return invoke<CreateHDWalletResponse>("import_hd_wallet", {
    name,
    mnemonic,
    chains,
    password,
  });
}

// =============================================================================
// Address Derivation
// =============================================================================

/**
 * Derive a new address for an existing wallet
 */
export async function deriveWalletAddress(
  walletId: string,
  chainId: string,
  account: number = 0,
  index: number = 0
): Promise<DerivedAddress> {
  return invoke<DerivedAddress>("derive_wallet_address", {
    walletId,
    chainId,
    account,
    index,
  });
}

// =============================================================================
// Session Management
// =============================================================================

/**
 * Check if the wallet session is currently unlocked
 */
export async function isWalletUnlocked(): Promise<boolean> {
  return invoke<boolean>("is_wallet_unlocked");
}

/**
 * Lock the wallet (clear cached secrets from memory)
 */
export async function lockWallet(): Promise<void> {
  return invoke<void>("lock_wallet");
}

/**
 * Unlock a wallet with password
 */
export async function unlockWallet(
  walletId: string,
  password: string
): Promise<void> {
  return invoke<void>("unlock_wallet", { walletId, password });
}

File: apps/desktop/src-tauri/src/commands/wallet.rs
//! Tauri wallet commands
//!
//! These commands are exposed to the frontend via Tauri IPC.
//! All commands use the global WalletManager instance.

use std::sync::OnceLock;

use tauri::State;

use crate::wallet::core::WalletManager;
use crate::wallet::registry::ChainInfo;
use crate::wallet::types::{
    CreateHDWalletRequest, CreateHDWalletResponse, DerivedAddress,
    ImportHDWalletRequest, ValidateMnemonicResponse,
};

/// Global wallet manager instance
static WALLET_MANAGER: OnceLock<WalletManager> = OnceLock::new();

/// Get or initialize the global wallet manager
pub fn get_wallet_manager() -> &'static WalletManager {
    WALLET_MANAGER.get_or_init(WalletManager::new)
}

// =============================================================================
// Chain Information Commands
// =============================================================================

/// Get all supported blockchain chains
#[tauri::command]
pub fn get_supported_chains() -> Vec<ChainInfo> {
    get_wallet_manager().get_supported_chains()
}

/// Get only mainnet chains
#[tauri::command]
pub fn get_mainnet_chains() -> Vec<ChainInfo> {
    get_wallet_manager().get_mainnet_chains()
}

/// Validate an address for a specific chain
#[tauri::command]
pub fn validate_chain_address(chain_id: String, address: String) -> Result<bool, String> {
    get_wallet_manager()
        .validate_address(&chain_id, &address)
        .map_err(|e| e.to_string())
}

// =============================================================================
// Mnemonic Commands
// =============================================================================

/// Generate a new random mnemonic
#[tauri::command]
pub fn generate_mnemonic(word_count: usize) -> Result<String, String> {
    get_wallet_manager()
        .generate_mnemonic(word_count)
        .map(|m| m.as_str().to_string())
        .map_err(|e| e.to_string())
}

/// Validate an existing mnemonic phrase
#[tauri::command]
pub fn validate_mnemonic(phrase: String) -> ValidateMnemonicResponse {
    get_wallet_manager().validate_mnemonic(&phrase)
}

// =============================================================================
// Wallet Creation Commands
// =============================================================================

/// Create a new HD wallet
///
/// Generates a new mnemonic, derives addresses for specified chains,
/// and returns the mnemonic (for backup) along with the addresses.
///
/// # Security Note
/// The mnemonic is returned ONCE for backup purposes.
/// The frontend should display it for user to write down,
/// then prompt to verify backup before proceeding.
#[tauri::command]
pub fn create_hd_wallet(
    name: String,
    chains: Vec<String>,
    word_count: Option<usize>,
    password: String,
) -> Result<CreateHDWalletResponse, String> {
    let request = CreateHDWalletRequest {
        name,
        chains,
        word_count: word_count.unwrap_or(12),
    };

    get_wallet_manager()
        .create_hd_wallet(&request, &password)
        .map_err(|e| e.to_string())
}

/// Import an existing HD wallet from mnemonic
#[tauri::command]
pub fn import_hd_wallet(
    name: String,
    mnemonic: String,
    chains: Vec<String>,
    password: String,
) -> Result<CreateHDWalletResponse, String> {
    get_wallet_manager()
        .import_hd_wallet(&name, &mnemonic, &chains, &password)
        .map_err(|e| e.to_string())
}

// =============================================================================
// Address Derivation Commands
// =============================================================================

/// Derive a new address for an existing wallet
#[tauri::command]
pub fn derive_wallet_address(
    wallet_id: String,
    chain_id: String,
    account: Option<u32>,
    index: u32,
) -> Result<DerivedAddress, String> {
    get_wallet_manager()
        .derive_address(&wallet_id, &chain_id, account.unwrap_or(0), index)
        .map_err(|e| e.to_string())
}

// =============================================================================
// Session Management Commands
// =============================================================================

/// Check if the wallet session is unlocked
#[tauri::command]
pub fn is_wallet_unlocked() -> bool {
    get_wallet_manager().is_unlocked()
}

/// Lock the wallet (clear cached secrets)
#[tauri::command]
pub fn lock_wallet() {
    get_wallet_manager().lock();
}

/// Unlock a wallet with password
#[tauri::command]
pub fn unlock_wallet(wallet_id: String, password: String) -> Result<(), String> {
    get_wallet_manager()
        .unlock(&wallet_id, &password)
        .map_err(|e| e.to_string())
}
